### ISSUE TITLE
Property named '$type' is silently not migrated via Cosmos DB migration

### ISSUE THREAD

**Issue by nmehlei** [AzureCosmosDB/data-migration-desktop-tool#97]
Hey, 
First of all, thank you for this amazing tool. I was successfully able to migrate data directly from and to Cosmos DB.

Though, for some reason, the property named `$type` was not copied with the rest of the data and was silently omitted. This value is used by the .NET's JSON library Newtonsoft.Json for complex data type handling. The omitted field led to changed data deserialization behavior, which in turn led to issues for my applications.

To clarify, this:
<img width="436" alt="Source" src="https://github.com/AzureCosmosDB/data-migration-desktop-tool/assets/2812989/c5c56e49-ff0b-43af-9e56-f4ff76990502">

was changed to this:
<img width="291" alt="Target" src="https://github.com/AzureCosmosDB/data-migration-desktop-tool/assets/2812989/ca457e23-3a26-4d81-9f78-030840bec257">

My guess would be that either this is the case for all fields with an `$` in it, or, since the tool itself seems to use Newtonsoft.Json, this usage inadvertently stops propagating that field. I would assume this is unintentional behavior, i.e. a bug. Though if it's indeed intended behavior, then I would suggest outputting a warning of some kind if this is detected as well as adding a note to the documentation.

For my testing, I used version 2.1.4 on Mac OS X Ventura 13.0.

**Comment by Clapla** [comment]
I had the same problem happening. This could have been really dangerous for us. Is it something we missed in the migrationsettings?

**Comment by dperez83** [comment]
Had the very same problem with the property named $type and it took us hours to understand what was happening because the lack of the field.

**VERY VERY DANGEROUS BEHAVIOR!**

**Comment by JohnDStrasz** [comment]
Same problem. This tool would be great if this CRITICAL issue would be resolved.

**Comment by dperez83** [comment]
> Same problem. This tool would be great if this CRITICAL issue would be resolved.

Yes, it silently ignore and without warnings

**Comment by crpietschmann** [comment]
The problem:
It looks like by default Newtonsoft.Json is ignoring the '$type' property.

Troubleshooting:
I tried look into the code a little bit for this. If you tell Newtonsoft's `JsonConvert.Deserialize` to handle the type properties via `TypeNameHandling.All` then it attempts to deserialize the object as the type specified in the '$type' property, which will result in an exception since the library isn't there.

Initial Solution Thoughts:
It looks to me that the code for the extensions will need to be modified to use a custom JsonConverter that deserializes the Json to a `Dictionary<string, object>` or something that's compatible with the rest of the code. It appears to me that all the extensions are handling Json the same way, at least the Json and Cosmos extensions from my checking, so we'll likely want to create our own `JsonConverter` abstraction class that handles serialization and deserialization correctly, so that any extensions that need to can use it.

**Comment by bowencode** [comment]
Can anyone seeing this issue provide a reproduction scenario: source and sink used, settings, a sample of data used? In running some basic tests of JSON->Cosmos-NoSQL and Cosmos-NoSQL->JSON I didn't see the $type property being dropped so there's something specific about these scenarios that I'm missing.

I tested with this data as "json-in_typed.json":

```json
[
  {
    "$type": "System.Object",
    "RealEstateType": "apartmentRent",
    "Title": "Test221",
    "MatterportUrl": ""
  }
]
```

migrationsettings.json for import to Cosmos, including two different options for WriteMode that have different internal serialization behavior:

```json
{
  "Source": "json",
  "Sink": "cosmos-nosql",
  "SourceSettings": {
    "FilePath": "json-in_typed.json"
  },
  "SinkSettings": {
    "ConnectionString": "FILL IN",
    "Database": "database",
    "Container": "typed",
    "PartitionKeyPath": "/id",
    "RecreateContainer": true,
    "WriteMode": "InsertStream"
    //"WriteMode": "Insert"
  }
}
```

migrationsettings.json for export from Cosmos:

```json
{
  "Source": "cosmos-nosql",
  "Sink": "json",
  "SourceSettings": {
    "ConnectionString": "FILL IN",
    //"IncludeMetadataFields": true,
    "Database": "database",
    "Container": "typed"
  },
  "SinkSettings": {
    "FilePath": "json-out_typed.json",
    "Indented": true
  }
}
```

@JefSchraag: I assume your problem with "_" was on export from Cosmos, which excludes metadata properties (those with leading underscores) by default. The IncludeMetadataFields setting above changes that behavior.

**Comment by JohnDStrasz** [comment]
For our issue, it happens when the $type was a sub-property

{
    "id": "600004",
    "itemType": {
        "$type": "Test1.Domain.Test2.Types.ABC, Test1.Domain",
        "property1": false,
        "property2": false,
        "value": 3,
        "displayName": "ABC"
    },

We have the IncludeMetadataFields property set to true on the Source and Sink settings.

Our migrations picked up the _ properties of CosmosDb (e.g. _ts) at the top level once we started including the meta data flag.

Our use cases for the tool will be Cosmos-to-Cosmos, Cosmos-to-JSON and JSON-to-Cosmos transfers.

Let me know if there is anything else relevant to include to help with the investigation.

**Comment by bowencode** [comment]
@JohnDStrasz can you confirm whether #122 fixes this for you? I included the sample you provided in my testing and confirmed the change during export but it would be good to make sure there's not something else you're seeing.

**Comment by JohnDStrasz** [comment]
Apologies, how would I go about testing this change?

**Comment by JohnDStrasz** [comment]
> @JohnDStrasz can you confirm whether #122 fixes this for you? I included the sample you provided in my testing and confirmed the change during export but it would be good to make sure there's not something else you're seeing.

I found some more time to test and I debugged into the code. All Cosmos transfers fail, regardless of whether they have the $type property or not, in the UpsertItemAsync line of the PopulateItem method in the CosmosDataSinkExtension.cs file

```
 case DataWriteMode.Upsert:
     var upsertResponse = await container.UpsertItemAsync(item, cancellationToken: cancellationToken);
     statusCode = upsertResponse.StatusCode;
     break;
```

I am compiling in Debug mode.

**Comment by bowencode** [comment]
@JohnDStrasz Another recent change caused a conflict in the serializer settings which is now fixed so this should be back to working for round trip transfers to and from Cosmos.

**Comment by JohnDStrasz** [comment]
> @JohnDStrasz Another recent change caused a conflict in the serializer settings which is now fixed so this should be back to working for round trip transfers to and from Cosmos.

It works!!! I compared a doc between the source and destination and it was exact, except for the usual Cosmos internal fields.

This is huge for us John. Thank you so much. Looking forward to it being merged to master soon.

**Comment by bowencode** [comment]
Fixes are now included in Release 2.1.5. Thank you @JohnDStrasz for help reproducing and validating!
